garbage_collector/src/objectstore/lister.rs

```
@@ -15,7 +17,32 @@ pub(crate) async fn perform(
     // sleep poll interval to avoid issues with immediately polling the object store at startup
     info!("object store polling will start in {sleep_interval_minutes} configured minutes");
     sleep(Duration::from_secs(60 * sleep_interval_minutes)).await;
-    let mut items = object_store.list(None).await.context(ListingSnafu)?;
+
+    let ret = Backoff::new(&BackoffConfig::default())
+        .retry_with_backoff("list_os_files", || async {
+            select! {
+            _ = shutdown.cancelled() => {
+                    ControlFlow::Break(None)
+            },
+
+            ret = object_store.list(None) => {
+                    match ret {
+                        Ok(b) => ControlFlow::Break(Some(b)),
+                        Err(e) => ControlFlow::Continue(e),
+                    }
+                }
+            }
+        })
+        .await;
+
+    let ret = match ret {
+        Ok(s) => s,
+        Err(_) => return Ok(()), // unreachable, unless backoff configured with a deadline
+    };
+    let mut items = match ret {
+        Some(s) => s,
+        None => return Ok(()), // shutdown requested while trying to get first OS list
+    };
```

I'm sure there's a cleaner way to unwrap like this twice. Rust suggestions welcome!

---

If you wanted to keep the same semantics here (No items/Error returns `Ok(())`), then you could use the `unwrap_or()` method on the first result like this:

```rust
    // backoff retry to avoid issues with immediately polling the object store at startup
    let mut items = match Backoff::new(&BackoffConfig::default())
        .retry_with_backoff(
            // Do stuff
        )
        .await
        .unwrap_or(None)
    {
        Some(s) => s,
        None => return Ok(()),
    };
```

I'd consider maybe avoiding the silent error swallowing and instead let the caller deal with matching/mapping the error (they'll have to anyway as the function returns a `Result<T,E>`. In which case you could do something like this:

```rust
    let mut items = Backoff::new(&BackoffConfig::default())
        .retry_with_backoff(
            // Do stuff
        )
        .await
        .map_err(|_| Error::BackoffTimedOut)?
        .ok_or(Error::EmptyResult)?;
```

Ideally you'd probably want to source `BackoffTimedOut` from the value given to the `map_err()` closure, but I couldn't work out quickly what generic parameter to give the `BackoffError<T>`.

---

I struggled some that `retry_with_backoff ` returns a custom Result type where the Error type has to be an enum type which only has a single element, namely `BackoffTimedOut`. A better api of the library should allow the lambda/user function to return any error it wants. E.g. the user of `retry_with_backoff` wants to retry some errors but for other errors it wants to stop retry and pass that error up. That's not allowed right now. Basically all our iox code uses `retry_all_errors` so that hasn't seen a need. Dom's suggestion in a comment above to select on the retry and the shutdown is clearly and removes the need for the retry lambda to have different errors differently but I think that's the generic problem to solve (say we wanted to retry only some of the object store listing errors - e.g. "bucket does not exist" is not likely to be better on a retry.) #over engineering.

Thanks for the help here! I forgot about `unwrap_or` and TIL `map_err`.